File: power_dist.js
const NODES_DATA = [
    { id:'Inverter', r: 50},
    { id:'PV',       r: 40},
    { id:'Battery',  r: 40},
    { id:'Grid',     r: 40},
    { id:'Home',     r: 50},
    { id:'Heatpump', r: 40},
    { id:'EV',       r: 40},
    { id:'Rest',     r: 40},
];
const EDGES_DATA = [
    { value: 10, source: 'PV', target: 'Inverter' },
    { value: -2, source: 'Battery', target: 'Inverter' },
    { value: 0, source: 'Grid', target: 'Inverter' },
    { value: 8, source: 'Inverter', target: 'Home' },
    { value: 1, source: 'Home', target: 'Heatpump' },
    { value: 1, source: 'Home', target: 'Rest' },
    { value: 6, source: 'Home', target: 'EV' },
];
const PARAMS = {
    width: 600,
    height: 600,
    padding: 50,
    min: 0,
    max: 10,
    forceManyBodyStrength: -2500,
    forceLinkDistance: 80,
    unit: "",
    digits: 2
};

function configureJson(nodes_data_json, edges_data_json) {
    var nodes_data = JSON.parse(nodes_data_json);
    var edges_data = JSON.parse(edges_data_json);
    configure(nodes_data, edges_data);
}
function configure(nodes_data, edges_data) {
    d3.select('svg').classed('container', true).selectAll(".marker").interrupt();
    d3.select('svg').classed('container', true).selectAll("*").remove();

    var simulation = d3.forceSimulation()
        .force("link", d3.forceLink().id(function(d) { return d.id; }))
        .force("charge", d3.forceManyBody().strength(PARAMS.forceManyBodyStrength))
        .force("center", d3.forceCenter(PARAMS.width / 2, PARAMS.height / 2));

    const container = d3.select('svg').classed('container', true)
        .attr("width", PARAMS.width)
        .attr("height", PARAMS.height);

    const nodes = container
        .selectAll(".node")
        .data(nodes_data)
        .enter()
        .append("circle")
        .classed("node", true)
        .attr("r", node => node.r)

    const edges = container
        .selectAll(".edge")
        .data(edges_data)
        .enter()
        .append("line")
        .classed("edge", true)

    const edgesvalue = container
        .selectAll(".edgevalue")
        .data(edges_data)
        .enter()
        .append("text")
        .classed("edgevalue", true)
        .text(edge => {
            if(edge.value == undefined) return "---";
            return edge.value + " " + PARAMS.unit
        });

    const markers = container
        .selectAll(".marker")
        .data(edges_data)
        .enter()
        .append("circle")
        .classed("marker", true)

    const nodenames = container
        .selectAll(".nodetext")
        .data(nodes_data)
        .enter()
        .append("text")
        .classed("nodetext", true)
        .text(node => node.id);

    simulation
        .nodes(nodes_data)
        .on("tick", ticked)
    simulation.force("link")
        .links(edges_data)
        .distance(PARAMS.forceLinkDistance);

    function ticked() {
            edges
                .attr("x1", d => d.source.x + d.source.r * Math.cos(Math.atan2(d.source.y - d.target.y, d.source.x - d.target.x)+Math.PI))
                .attr("y1", d => d.source.y + d.source.r * Math.sin(Math.atan2(d.source.y - d.target.y, d.source.x - d.target.x)+Math.PI))
                .attr("x2", d => d.target.x + d.target.r * Math.cos(Math.atan2(d.source.y - d.target.y, d.source.x - d.target.x)))
                .attr("y2", d => d.target.y + d.target.r * Math.sin(Math.atan2(d.source.y - d.target.y, d.source.x - d.target.x)));
            nodes
                .attr("cx", d => d.x)
                .attr("cy", d => d.y);
            nodenames
                .attr("x", d => d.x)
                .attr("y", d => d.y);
            edgesvalue
                .attr("x", d => (d.source.x + d.target.x)/2)
                .attr("y", d => (d.source.y + d.target.y)/2);
        }
}

function setParam(key, value) {
    PARAMS[key] = value;
}

function animate() {
    const container = d3.select('svg').classed('container', true);
    const markers = container.selectAll(".marker");
    const edges = container.selectAll(".edge")

    markers.each(function transition(marker) {
        d3.select(this)
        .transition()
        .duration((d) => {
            const edgesvalue = container.selectAll(".edgevalue");
            const value = d.value == undefined ? 0 : d.value;
            
            if(d.value == undefined)
                edgesvalue.nodes()[d.index].textContent = "---";
            else
                edgesvalue.nodes()[d.index].textContent = value + " " + PARAMS.unit;

            const norm = (Math.min(Math.max(Math.abs(value), PARAMS.min), PARAMS.max) - PARAMS.min) / (PARAMS.max - PARAMS.min);
            const duration = 2500 - norm * 2000;

            return duration
        })
        .attrTween("cx", (d) => {
            return (t) => {
                const value = d.value == undefined ? 0 : d.value;
                const edge = edges.nodes()[d.index];
                const totalLen = edge.getTotalLength();
                const pointIdx = value == 0 ? 0 : (value > 0  ? totalLen*t : totalLen*(1-t));
                const point = edge.getPointAtLength(pointIdx);
                return point.x;
            }; 
        })
        .attrTween("cy", (d) => { 
            return (t) => {
                const value = d.value == undefined ? 0 : d.value;
                const edge = edges.nodes()[d.index];
                const totalLen = edge.getTotalLength();
                const pointIdx = value == 0 ? 0 : (value > 0  ? totalLen*t : totalLen*(1-t));
                const point = edge.getPointAtLength(pointIdx);
                return point.y;
            }; 
        })
        .on("end", transition);
    });
}

function configureDefault() {
    configure(NODES_DATA, EDGES_DATA);
    animate();
}

function valueChanged(jsonData) {
    var input = JSON.parse(jsonData);    

    const container = d3.select('svg').classed('container', true);
    const edges = container.selectAll(".edge");
    const data = edges.data();

    data.forEach(d => {
        d.value = undefined;
    });

    input.Variables.forEach(function(variable) {
        source = variable.UniqueName.split("->")[0];
        target = variable.UniqueName.split("->")[1];
        
        entry = data.find(d => d.source.id==source && d.target.id==target);
        if(entry === undefined) return;
        if(variable.Value == "---") return;
        entry.value = parseFloat(variable.Value).toFixed(PARAMS.digits);
    });
}

File: power_dist.py
# Available global variables:

# Application - The Application object.
# Instrument - The Instrument object.

from win32com.client import Dispatch, DispatchWithEvents
import random, json

EXAMPLE_TERMS = ["Robot", "Conveyor", "PLC", "Scanner", "Actuator", "Cobot", "AGV", "Sensor", "HMI", "Servo", "CNC", 
                 "Manipulator", "Actuator", "Stepper", "VFD", "Gripper", "DCS", "Inverter", "Encoder", "Feeder", 
                 "Engine", "ECU", "ABS", "Airbag", "Turbo", "Differential", "Injector", "Transmission", "Hybrid", 
                 "Throttle", "Steering", "CVT", "Catalytic", "OBD", "TPMS", "Cruise", "Ignition", "Suspension", 
                 "Radiator", "Fuel pump", "Lathe", "Milling", "Router", "Oscilloscope", "Multimeter", "Press", 
                 "Gearbox", "Motor", "Pump", "Exchanger", "Oven", "Fan", "3D Printer", "Welder", "Solder", "Cutter", 
                 "Conveyor", "Generator", "HVAC", "Compressor", "Solar", "Wind", "Battery", "Inverter", "Rainwater", 
                 "Geothermal", "Charger", "Composter", "Gasifier", "Thermostat", "Lighting", "Controls", "Heater",
                 "Turbine", "Emission", "Digester", "Ventilator", "Irrigation", "Purifier", "Composter"]

class InstrumentEvents(object):
    def __init__(self):
        sender = Dispatch(Instrument)
        
        # update HTML/JS
        nodes_data = [{"id": cp.Value, "r": 40} for cp in sender.CustomProperties if cp.Name == "Node"]
        edges_data = [{"value": 0, "source":cp.Value.split("->")[0], "target":cp.Value.split("->")[1]} for cp in sender.CustomProperties if cp.Name == "Link" and cp.Value.split("->")[0] in [n["id"] for n in nodes_data] and cp.Value.split("->")[1] in [n["id"] for n in nodes_data]]
        cps = { cp.Name:cp for cp in sender.CustomProperties}
        
        for cpName in ["Unit", "Digits", "Min", "Max", "Width", "Height", "Padding"]:
            Instrument.InvokeScript('setParam', (cps[cpName].Name.lower(), cps[cpName].Value))        
        Instrument.InvokeScript('configureJson', (json.dumps(nodes_data), json.dumps(edges_data), ))
        Instrument.InvokeScript('animate', ())
        
    def OnCustomPropertyChanged(self, sender_, customProperty_, action):
        if action != 0: return

        customProperty = Dispatch(customProperty_)
        sender = Dispatch(sender_)
        
        #
        # Add and remove "Nodes" and "Links"
        #
        nodesCp = list(filter(lambda cp: cp.Name == "Node", sender.CustomProperties))
        linksCp = list(filter(lambda cp: cp.Name == "Link", sender.CustomProperties))
            
        if customProperty.Name == 'Number of Nodes':
            customProperty.Value = 0 if customProperty.Value < 0 else customProperty.Value
            customProperty.Value = len(EXAMPLE_TERMS) if customProperty.Value > len(EXAMPLE_TERMS) else customProperty.Value
            
            diff = customProperty.Value - len(nodesCp)
            if diff > 0: # add Node CPs
                for idx in range(0, diff):
                    prop = sender.CustomProperties.Add(1, "Node")
                    prop.Category = "Power Distribution Nodes"
                    # find default node value
                    usedValues = [cp.Value for cp in nodesCp]
                    nameCandidate = ""
                    while nameCandidate == "" or nameCandidate in usedValues: # avoid same names
                        nameCandidate = random.choice(EXAMPLE_TERMS)
                    prop.Value = nameCandidate
                    
            elif diff < 0: # remove Node CPs
                for idx in range(0, abs(diff)):
                    idx = [cp for cp in sender.CustomProperties].index(nodesCp.pop())
                    sender.CustomProperties.Delete(idx)
                   
        elif customProperty.Name == 'Number of Links':
            customProperty.Value = 0 if customProperty.Value < 0 or len(nodesCp) == 0 else customProperty.Value
            
            diff = customProperty.Value - len(linksCp)
            if diff > 0: # add Link CPs
                for idx in range(0, diff):
                    prop = sender.CustomProperties.Add(1, "Link")
                    prop.Category = "Power Distribution Links"
                    
                    # find default link value
                    usedValues = [cp.Value for cp in list(filter(lambda cp: cp.Name == "Link", sender.CustomProperties))]
                    nameCandidate = ""
                    while nameCandidate == "" or nameCandidate in usedValues:
                        nameCandidate = random.choice(EXAMPLE_TERMS) + "->" + random.choice(EXAMPLE_TERMS)
                        usedValues = [cp.Value for cp in list(filter(lambda cp: cp.Name == "Link", sender.CustomProperties))]
                    prop.Value = nameCandidate
            elif diff < 0: # remove Link CPs
                for idx in range(0, abs(diff)):
                    idx = [cp for cp in sender.CustomProperties].index(linksCp.pop())
                    sender.CustomProperties.Delete(idx)

        elif customProperty.Name == 'Node':
            nodeProps = [cp for cp in sender.CustomProperties if cp.Name == "Node"]
            nodeProps.remove(customProperty)
            nodeNames = [np.Value for np in nodeProps]
            while customProperty.Value in nodeNames: # avoid same names
                customProperty.Value = random.choice(EXAMPLE_TERMS)
        else:
            Instrument.InvokeScript('setParam', (customProperty.Name.lower(), customProperty.Value))
        
        #
        # Sync ConnectionNodes with "Link"-custom properties
        #
        linkProps = [cp for cp in sender.CustomProperties if cp.Name == "Link"]
        connNodes = {cn.UniqueName: cn for cn in sender.ConnectionNodes}

        linkPropsNames = [np.Value for np in linkProps]
        connNodesNames = [key for key in connNodes]
        
        toAdd = set(linkPropsNames).difference(set(connNodesNames))
        toRemove = set(connNodesNames).difference(set(linkPropsNames))
        
        for cnName in toAdd:
            cn = sender.ConnectionNodes.Add()
            cn.UniqueName = cnName
        for cnName in toRemove:
            connNodes[cnName].Delete()
        
        # update HTML/JS
        nodes_data = [{"id": cp.Value, "r": 40} for cp in sender.CustomProperties if cp.Name == "Node"]
        edges_data = [{"value": 0, "source":cp.Value.split("->")[0], "target":cp.Value.split("->")[1]} for cp in sender.CustomProperties if cp.Name == "Link" and cp.Value.split("->")[0] in [n["id"] for n in nodes_data] and cp.Value.split("->")[1] in [n["id"] for n in nodes_data]]
        
        Instrument.InvokeScript('configureJson', (json.dumps(nodes_data), json.dumps(edges_data), ))
        Instrument.InvokeScript('animate', ())

    def OnConnectionAdded(self, sender, connectionObject, propertyName, variable):
        pass
    def OnConnectionAdding(self, sender, connectionObject, propertyName, variable, connectionAddingEventArgument):
        pass
    def OnConnectionDeleted(self, sender, connectionObject, propertyName, variable):
        pass
    def OnConnectionDeleting(self, sender, connectionObject, propertyName, variable):
        pass
    def OnDataReceived(self, sender, key, value):
        pass
    def OnDeleting(self, sender):
        pass
    def OnKeyDown(self, sender, key, modifierKeys):
        pass
    def OnKeyPress(self, sender, keyChar):
        pass
    def OnKeyUp(self, sender, key, modifierKeys):
        pass
    def OnMouseClick(self, sender, mouseButton, x, y, modifierKeys):
        pass
    def OnMouseDoubleClick(self, sender, mouseButton, x, y, modifierKeys):
        pass
    def OnMouseDown(self, sender, mouseButton, x, y, modifierKeys):
        pass
    def OnMouseEnter(self, sender):
        pass
    def OnMouseHover(self, sender):
        pass
    def OnMouseLeave(self, sender):
        pass
    def OnMouseMove(self, sender, x, y, modifierKeys):
        pass
    def OnMouseUp(self, sender, mouseButton, x, y, modifierKeys):
        pass
    def OnMouseWheel(self, sender, delta, modifierKeys):
        pass
    def OnSelectionStateChanged(self, sender, isSelected):
        pass
    def OnUrlChanged(self, sender, url):
        pass
    def OnValidStateChanged(self, sender, connectionObject, propertyName, variable, isValid):
        pass

InstrumentWithEvents = DispatchWithEvents(Instrument, InstrumentEvents)
